src/v/coproc/script_context_backend.cc

```diff
@@ -189,69 +192,126 @@ process_one_reply(process_batch_reply::data e, output_write_args args) {
+    auto found = args.inputs.find(source);
+    if (found == args.inputs.end()) {
+        throw bad_reply_exception(
+          fmt::format("script {} unknown source ntp: {}", args.id, source));
+    }
+    auto src_ptr = found->second;
+    for (auto& e : reply_group) {
+        try {
+            co_await process_one_reply(std::move(e), src_ptr, args);
+        } catch (const bad_reply_exception& ex) {
+            /// A bad reply shouldn't affect the outcome, just ignore
+            vlog(coproclog.error, "Erraneous response detected: {}", ex);
+        }
+    }
+    /// If we've reached here without error, all materialized
+    /// partitions (working under the input) can be promoted, rational is to
+    /// support filter operations, i.e. responses were returned for some
+    /// materialized partitions but not others. If offsets for missing ones
+    /// were not bumped in case of success, there would be an infinite retry
+    /// loop.
+    for (auto& [_, o] : src_ptr->wctx.offsets) {
+        if (o <= src_ptr->rctx.last_read) {
+            /// Omit increasing offets that are ahead of current read, only
+            /// possibility of this case would be a load of a stored offset that
+            /// is far ahead
+            o = src_ptr->rctx.last_read;
+        }
     }
-    auto ofound = ntp_ctx->offsets.find(args.id);
-    vassert(
-      ofound != ntp_ctx->offsets.end(),
-      "Offset not found for script id {} for ntp owning context: {}",
-      args.id,
-      ntp_ctx->ntp());
-    /// Reset the acked offset so that progress can be made
-    ofound->second.last_acked = ofound->second.last_read;
 }
 
 ss::future<>
 write_materialized(output_write_inputs replies, output_write_args args) {
     if (replies.empty()) {
         vlog(
           coproclog.error, "Wasm engine interpreted the request as erraneous");
-    } else {
-        for (auto& e : replies) {
-            co_await process_one_reply(std::move(e), args);
-        }
+        co_return;
+    }
+    grouping_t grs = group_replies(std::move(replies));
+    bool err{false};
+    co_await ss::parallel_for_each(
+      grs, [args, &err](grouping_t::value_type& vt) -> ss::future<> {
+          try {
+              co_await process_reply_group(
+                vt.first, std::move(vt.second), args);
+          } catch (const bad_reply_exception& ex) {
+              vlog(coproclog.error, "No source for reply: {}", ex);
+          } catch (const script_failed_exception& ex) {
+              throw ex;
+          } catch (const coproc::exception& ex) {
+              /// For any type of failure the offset will not be touched,
+              /// the read phase will always read from the global min of all
+              /// offsets ever registered.
+              vlog(coproclog.info, "Error while processing reply: {}", ex);
+              err = true;
+          }
+      });
+    if (err) {
+        co_await ss::sleep(100ms);
```

Comment:
Should this be configurable?

Reply:
Not sure, its only for cases where retry will occur because some action which has already initiated hasn't completed. For example creating a topic cannot be waited on because topics are considered fully created only when the controller backend has added the metadata into the shard table and created the respective partition and/or log instance. 

I previously had some code to wait on these events but it gets hairy and not sure if its worth the extra complexity. So to answer the question i didn't make this configurable because this is one of those events that should resolve itself soon-ish.